### PR TITLE
refactor: package addons

### PR DESCRIPTION
**Pull request by dannyrandall**
Refactor based on feedback in a previous PR: https://github.com/aws/copilot-cli/pull/3820#discussion_r931683092

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/addon/addons.go [hunk 1]
```diff
@@ -68,54 +58,75 @@ func New(wlName string) (*Addons, error) {
 	}, nil
 }
 
-// NewPackager creates an Addons struct that will package local artifacts when
-// generating the addons template.
-// See https://docs.aws.amazon.com/cli/latest/reference/cloudformation/package.html for more details.
-func NewPackager(wlName string, bucket string, uploader uploader) (*Addons, error) {
-	addons, err := New(wlName)
+type Stack struct {
```
Comment (efekarakus) [inline internal/pkg/addon/addons.go:39]: doc comments 🥺 (they're missing from methods and exported types)

File: internal/pkg/addon/addons.go [hunk 2]
```diff
@@ -68,54 +58,75 @@ func New(wlName string) (*Addons, error) {
...
+	template   *cfnTemplate
+	parameters yaml.Node
+	wlName     string
+}
+
+func (a *Addons) Stack() (*Stack, error) {
```
Comment (efekarakus) [inline internal/pkg/addon/addons.go]: I wonder if we need the method 🤔 it seems like clients are always forced to do 2-steps:
```go
addons := addons.New(/* params... */)
stack, err := addons.Stack()
```
Can we get rid of the `New` constructor and `Stack` method and replace it with a single `Parse` function:
```go
// Parse reads the addons/ directory under the workloadName and returns a stack configuration.
func Parse(workloadName string, ws WorkspaceReader) (*Stack, error) {
  // New + addons.Stack()
}
```

File: internal/pkg/addon/package.go [hunk 3]
```diff
@@ -190,14 +190,27 @@ var resourcePackageConfig = map[string][]packagePropertyConfig{
...
+	Uploader      uploader
+	WorkspacePath string
+	Fs            *afero.Afero
+
+	wlName string
+}
+
+// Package finds references to local files in Stack's template, uploads
+// the files to S3, and replaces the file path with the S3 location.
+func (s *Stack) Package(p PackageConfig) error {
```
Comment (efekarakus) [inline internal/pkg/addon/package.go]: smh 
```suggestion
func (s *Stack) Package(cfg PackageConfig) error {
```

File: internal/pkg/addon/package.go [hunk 4]
```diff
@@ -190,14 +190,27 @@ var resourcePackageConfig = map[string][]packagePropertyConfig{
 	},
 }
 
-func (t *cfnTemplate) pkg(a *Addons) error {
-	err := a.packageIncludeTransforms(&t.Metadata, &t.Mappings, &t.Conditions, &t.Transform, &t.Resources, &t.Outputs)
+type PackageConfig struct {
+	Bucket        string
+	Uploader      uploader
+	WorkspacePath string
+	Fs            *afero.Afero
```
Comment (efekarakus) [inline internal/pkg/addon/package.go]: nit: can we depend on the `afero.Fs` interface instead?
```go
FS      afero.Fs // all caps like io.FS
```

File: internal/pkg/addon/package.go [hunk 5]
```diff
@@ -288,7 +301,7 @@ func yamlMapGet(node *yaml.Node, key string) *yaml.Node {
 	return &yaml.Node{}
 }
 
-func (a *Addons) packageProperty(resourceProperties *yaml.Node, pkgCfg packagePropertyConfig) error {
+func (p *PackageConfig) packageProperty(resourceProperties *yaml.Node, pkgCfg packagePropertyConfig) error {
```
Comment (efekarakus) [inline internal/pkg/addon/package.go]: nit:
```suggestion
func (pkg *PackageConfig) packageProperty(resourceProperties *yaml.Node, propCfg packagePropertyConfig) error {
```

File: internal/pkg/cli/deploy/svc.go [hunk 6]
```diff
@@ -969,26 +976,36 @@ func (d *workloadDeployer) pushEnvFileToS3Bucket(in *pushEnvFilesToS3BucketInput
...
+	}
 
-func (d *workloadDeployer) pushAddonsTemplateToS3Bucket(in *pushAddonsTemplateToS3BucketInput) (string, error) {
-	tmpl, err := in.templater.Template()
-	if err != nil {
-		var notFoundErr *addon.ErrAddonsNotFound
-		if errors.As(err, &notFoundErr) {
-			// addons doesn't exist for service, the url is empty.
-			return "", nil
+	if false { // TODO remove to enable packaging addons
```
Comment (efekarakus) [inline internal/pkg/cli/deploy/svc.go:976]: lol 😂 should we remove it now? Isn't this the last milestone

**Review by efekarakus — COMMENTED** [review]
Looks great! just one additional refactor request regarding the `Stack`method and constructor

File: internal/pkg/cli/deploy/svc.go (hunk 6 above)
Comment (dannyrandall) [reply internal/pkg/cli/deploy/svc.go:976]: no 😤 this is a `refactor` pr. i'm very excited for my `feat` pr 😜

File: internal/pkg/addon/addons.go (hunk 2 above)
Comment (dannyrandall) [reply internal/pkg/addon/addons.go]: i like this idea! def should help clean it up

**Review by efekarakus — APPROVED** [review]
:shipit:

File: internal/pkg/addon/addons_test.go [hunk 7]
```diff
@@ -242,99 +185,87 @@ func TestAddons_Template(t *testing.T) {
...
+			stack, err := Parse(tc.workloadName, mocks.ws)
+			if tc.wantedErr != nil {
+				require.EqualError(t, err, tc.wantedErr.Error())
+				return
+			}
+			require.NoError(t, err)
 
-			// THEN
+			template, err := stack.Template()
 			if tc.wantedErr != nil {
```
Comment (Lou1415926) [inline internal/pkg/addon/addons_test.go]: I might have missed something - does this ever catch a `wantedErr` for `Template()` call? It seems like we are testing a combination of `Parse() + Stack()` - would the below snippets achieve what we want to test.
```go
	stack, err := Parse(tc.workloadName, mocks.ws)
	if tc.wantedErr != nil {
		require.EqualError(t, err, tc.wantedErr.Error())
		return
	}
	require.NoError(t, err)

         template, err := stack.Template()
         require.NoErr(err)
         require.Equal(t, tc.wantedTemplate, template)
```

File: internal/pkg/addon/testdata/merge/wanted.yaml [hunk 8]
```diff
@@ -1,146 +1,146 @@
 Metadata:
-    Version: 1
```
Comment (Lou1415926) [inline internal/pkg/addon/testdata/merge/wanted.yaml:2]: ah I'm curious which part of the codes is causing these changes? I think we've been using `setIndent(2)` all along :think_spin

**Review by Lou1415926 — APPROVED** [review]
Awesome thanks!! The code LGTM :shipit:  just some nits (more like questions)on testings

File: internal/pkg/addon/testdata/merge/wanted.yaml (hunk 8 above)
Comment (dannyrandall) [reply internal/pkg/addon/testdata/merge/wanted.yaml:2]: we weren't using it for `Template()`, but we were for `Parameters()` 😅

Comment (dannyrandall) [reply internal/pkg/addon/testdata/merge/wanted.yaml:2]: so yeah, it's bc we are using it in `Template()` now too

File: internal/pkg/addon/addons_test.go (hunk 7 above)
Comment (dannyrandall) [reply internal/pkg/addon/addons_test.go]: good catch! it wasn't doing anything there 😊

**Comment by dannyrandall** [comment]
@Mergifyio refresh

**Pull request merged**